{"nbformat":4,"nbformat_minor":0,"metadata":{"celltoolbar":"Slideshow","colab":{"name":"L11 Files.ipynb","provenance":[],"collapsed_sections":[]},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.7.4"}},"cells":[{"cell_type":"markdown","metadata":{"id":"5UyFJYXOx-WH"},"source":["# Lecture 11 - Files (https://bit.ly/39pWEle)\n","\n","Today:\n","* Files.\n","  * Writing Files\n","  * Reading files\n","  * With keyword\n","  * For loops on file handles\n","  * Binary files \n","  * Large worked example (Fasta file processing)\n","  * Directories\n","  * Web file retrieval\n","\n"]},{"cell_type":"markdown","metadata":{"id":"iT7epkfjCAq6"},"source":["# Motivation\n","\n","* Up to now we've mainly used input() to get input to the programs we've written.\n","\n","* In the real world, this is not the way we generally get data into a program. Rather we read and write data to and from files.\n","\n","* Files, like documents, or spread sheets or images, are just semantically organized arrays of bits. \n","\n","* They can exist on a disk, be pulled from the internet or be created transiently as we pipe inputs and outputs from our programs. \n","\n","* They can be small, and they can be very large. Typically the data associated with a scientific analysis will be large relative to the size of the program. For this reason we have to think about processing the data in stages using streaming. \n","\n","Anyway, let's dive in... (as an aside, the examples today will show larger code samples than we've seen previously - you're leveling up)"]},{"cell_type":"markdown","metadata":{"id":"SZ27fsOR87Cw"},"source":["# Writing Files\n","\n","Let's first look at creating a file. You'll often want to do this when writing out the results of some analysis."]},{"cell_type":"code","metadata":{"id":"38xJY3yOui52"},"source":["my_file = open(\"test.txt\", \"w\") # Call to open() opens a file\n","# test.txt in the working directory of the Python program.\n","# The \"w\" string argument (short for \"write\"), tells open()\n","# that you intend to write to the file. \n","\n","# If test.txt does not exist before this point it is created as an empty file.\n","\n","# If test.txt existed before this point its contents are now overwritten - \n","# so be careful opening files for writing\n","\n","# my_file (the return value from open) is a \"file handle\", that is you use it\n","# to handle the communication with the file\n","\n","\n","\n","# The write method of the file_handle object is a lot like print(), except\n","# the contents of the string arguments are written to the file, instead of the\n","# screen, and write does not add newlines to the end of the string (so you add them\n","# explicitly, as in this example)\n","my_file.write(\"My first file written from Python\\n\") \n","my_file.write(\"---------------------------------\\n\")\n","my_file.write(\"Hello, world!\\n\")\n","\n","my_file.close() # This closes the file. You can't keep using my_file after this\n","# as the file connection is closed. It is important to close the file when \n","# you're done to ensure the file writing is completed successfully (although Python\n","# will clean things up automatically on exit, you shouldn't rely on this in cases\n","# where you're program gets terminated prematurely)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"2cf01WIn7pJZ"},"source":["Note: (Do matched demo from terminal to show contents of file)\n"]},{"cell_type":"markdown","metadata":{"id":"KjJ_lZssTKgN"},"source":["**To append to the end of an existing file** "]},{"cell_type":"code","metadata":{"id":"2EpL3R5WBeVf"},"source":["my_file = open(\"test.txt\", \"a\") # The 'a' argument (short for append) opens an \n","# existing file to append to it \n","# If test.txt does not exist then 'a' creates the file, just like 'w'\n","\n","my_file.write(\"Hello, again!\\n\") \n","\n","my_file.close() # Again, always close the file"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"bf-ylIdLTQpE"},"source":["Note: (Continue matched demo from terminal to show contents of appended to file)"]},{"cell_type":"markdown","metadata":{"id":"0Csg1vPc8_7C"},"source":["# Reading Files"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":157},"id":"CAhQ4GB07xBM","executionInfo":{"elapsed":503,"status":"ok","timestamp":1572561573602,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"031dc175-94e2-4a44-e410-0e5eb5e2dfe2"},"source":["mynewhandle = open(\"test.txt\", \"r\") # This opens the file, \"r\" argument means \"read\"\n","\n","# The following loop walks through\n","# the file line by line\n","\n","while True:                            # Keep reading forever\n","  \n","    theline = mynewhandle.readline()   # Try to read next line using readline()\n","    \n","    if theline == \"\":              # When done, readline returns the empty string ''\n","      # allowing us to detect the end of the file and to leave the loop\n","      \n","        break                          #     leave the loop\n","\n","    # Now process the line we've just read\n","    print(theline), end=\"\")\n","\n","mynewhandle.close() # This might seem less important, as you're just reading \n","# from the file, however most underlying operating systems have limits on\n","# the number of active file handles, so it is good to cleanup. It is also \n","# generally a bad idea to have multiple file handles open on one file (unless they are\n","# all read only file handles)"],"execution_count":null,"outputs":[{"output_type":"stream","text":["My first file written from Python\n","\n","---------------------------------\n","\n","Hello, world!\n","\n","Hello, again!\n","\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"XpE5L7547yKT"},"source":["What is nice about the above loop is that regardless of the number of files, Python only stores one line in memory, allowing it to process arbitrarily large files.\n","\n","Python will complain if you try to open a file that doesn't exist..."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":166},"id":"aTvzYDew7yiC","executionInfo":{"elapsed":925,"status":"error","timestamp":1572561338378,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"94122b46-28f1-431b-a42b-b5351990a730"},"source":["mynewhandle = open(\"wharrah.txt\", \"r\")"],"execution_count":null,"outputs":[{"output_type":"error","ename":"FileNotFoundError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-5-90876b2f847b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mmynewhandle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"wharrah.txt\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"r\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'wharrah.txt'"]}]},{"cell_type":"markdown","metadata":{"id":"ooeE2DB6FCRM"},"source":["To read contents of file into memory in one go use read()"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":122},"id":"aGPsxgzYGuf1","executionInfo":{"elapsed":950,"status":"ok","timestamp":1572561373895,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"1578e898-5c5e-4634-8238-8135069427ae"},"source":["fh = open(\"test.txt\", \"r\") # I often use fh to denote a file handle, for some reason\n","\n","content = fh.read() # Read the contents of the file\n","\n","print(type(content)) # Yup, content is a string\n","\n","print(content)\n","\n","fh.close()"],"execution_count":null,"outputs":[{"output_type":"stream","text":["<class 'str'>\n","My first file written from Python\n","---------------------------------\n","Hello, world!\n","Hello, again!\n","\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"fUUcCij9H0Vi"},"source":["readlines will read all the lines in a file:"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":72},"id":"UyOOG16sHugL","executionInfo":{"elapsed":954,"status":"ok","timestamp":1572561427610,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"85bde637-6439-490f-cec9-0398d9bc293a"},"source":["fh = open(\"test.txt\") # I often use fh to denote a file handle, for some reason\n","# If you don't give the second argument it defaults to read only\n","\n","content = fh.readlines() # Read the contents of the file\n","\n","print(type(content)) # Yup, content is now a list of strings\n","\n","print(content)\n","\n","fh.close()"],"execution_count":null,"outputs":[{"output_type":"stream","text":["<class 'list'>\n","['My first file written from Python\\n', '---------------------------------\\n', 'Hello, world!\\n', 'Hello, again!\\n']\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"nq84RePQIrbm"},"source":["# The With Keyword\n","\n","Good programming is often about anticipating errors and dealing with them.\n","\n","The 'with' keyword is both useful short-hand and good defensive programming. We don't need to know the details (curious: https://docs.python.org/2.5/whatsnew/pep-343.html), but basically it cleans up the file handle for you"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":236},"id":"tZU5iNqQIqcF","executionInfo":{"elapsed":513,"status":"error","timestamp":1572561758891,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"3f0b7911-5255-493c-d0f5-ac5fa63030a3"},"source":["with open(\"test.txt\") as fh: # With removes the need for the \"fh.close()\" statement\n","  # and is better because fh.close() is guaranteed to be run even if there is an error\n","  # in processing the file\n","  \n","  assert False # This causes an error, but fh will still be closed\n","  \n","  content = fh.readlines() \n","  print(content)\n","  \n","  \n","# For files the syntax is:\n","# with OPEN_EXPRESSION as FILE_HANDLE:\n","#   STATEMENTS THAT USE FILE_HANDLE"],"execution_count":null,"outputs":[{"output_type":"error","ename":"AssertionError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)","\u001b[0;32m<ipython-input-10-2efbed2a78a5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m   \u001b[0;31m# in processing the file\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m   \u001b[0;32massert\u001b[0m \u001b[0;32mFalse\u001b[0m \u001b[0;31m# This causes an error, but fh will still be closed\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m   \u001b[0mcontent\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfh\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadlines\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mAssertionError\u001b[0m: "]}]},{"cell_type":"markdown","metadata":{"id":"Ji9VkG3QO2JH"},"source":["# For loop on file:"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":87},"id":"PbI6o-TXOtJ2","executionInfo":{"elapsed":869,"status":"ok","timestamp":1572561908305,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"1b95e776-e61a-4e34-d5c9-69f433f80672"},"source":["# If you run a for loop on a file handle open for reading then\n","# it defaults to reading lines, this is compact and memory efficient (because\n","# we only read a line at a time from the file)\n","\n","with open(\"test.txt\") as fh:\n","  for line in fh:\n","    print(line, end=\"\") # end=\"\" stops print() from adding a newline to the end"],"execution_count":null,"outputs":[{"output_type":"stream","text":["My first file written from Python\n","---------------------------------\n","Hello, world!\n","Hello, again!\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"YSP_EVs7QpgW"},"source":["# Binary Files\n","\n","All the files we've seen so far are assumed to be text files (i.e. composed of human readable stuff). However, Python can also happily process binary files, which are just files with arbitrary bits in them. \n","\n","On the left is a binary file - the bits aren't organized so that they can be decoded as text. We can't assume that the bits in a binary file are organized into bytes that represent text characters with white space, etc. We can always treat a text file as a binary file, it's just that we ignore its formatting.\n","\n","On the right is a text file - the bits are formatted so that each byte or 16 bit word represents a text character.\n","\n","![alt text](https://www.syncfusion.com/books/Java_Succinctly_Part_2/Images/binary-file-vs-text-file.png)\n"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":72},"id":"AzQsT6aPXhkj","executionInfo":{"elapsed":477,"status":"ok","timestamp":1572562141485,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"8f494076-51a6-404c-fb08-a54bee4a392c"},"source":["# Reading and writing binary data\n","\n","with open(\"test.txt\", \"rb\") as f: # Here we treat test.txt as a binary file - any file\n","  # can be considered just a collection of bits\n","  \n","  with open(\"test2.txt\", \"wb\") as g: # Note the nested with statements\n","\n","    while True:\n","      buf = f.read(1024) # The argument to read tells it to read a set number of \n","      # bytes - each byte is an 8 bit (8 0's or 1's) words into the \"buf\" object, \n","      # which is a string\n","\n","      #print(buf)\n","      \n","      if len(buf) == 0: # We're at the end\n","         break\n","          \n","      g.write(buf)"],"execution_count":null,"outputs":[{"output_type":"stream","text":["b'My first file written from Python\\n---------------------------------\\nHello, world!\\nHello, again!\\n'\n","b''\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"WgWDW4p-Y2Xb"},"source":["Now we can see that test2.txt is just a copy of the original file:"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":87},"id":"oRO95FSoY9MI","executionInfo":{"elapsed":867,"status":"ok","timestamp":1572562209509,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"75db7488-f6d9-4604-a510-77519a48b61e"},"source":["with open(\"test2.txt\") as fh:\n","  for line in fh:\n","    print(line, end=\"\") # end=\"\" stops print() from adding a newline to the end"],"execution_count":null,"outputs":[{"output_type":"stream","text":["My first file written from Python\n","---------------------------------\n","Hello, world!\n","Hello, again!\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"jZBKbIBn9Woa"},"source":["# File Processing Example\n","\n","Here's a fairly complete worked example of doing file writing and parsing using the simplest bioinformatics format, the fasta file.\n","\n","A fasta file is a file that stores nucleotide (like DNA, RNA,) and amino-acid (protein) sequences.\n","\n","You should be able to understand all this code and what it is doing. If you don't, play with it until you do!\n","\n","\n"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":282},"id":"2YPQzixw8B1q","executionInfo":{"elapsed":880,"status":"ok","timestamp":1572562678463,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"e2d5bde6-702d-4ca5-bca4-373e4ad1591b"},"source":["\"\"\"\n","Example creating read and write methods for fasta files.\n","\n","Fasta format is a dead simple text (ascii) file:\n","\n",">HEADER_1\n","SEQUENCE_LINE_1\n","SEQUENCE_LINE_2\n","...\n","SEQUENCE_LINE_N\n",">HEADER_2\n","SEQUENCE_LINE_1\n","SEQUENCE_LINE_2\n","...\n","SEQUENCE_LINE_N\n","\n","Where HEADER_ lines give a string descriptor of the sequence\n","and the SEQUENCE_LINE_ lines are concatenated together to form the actual \n","amino-acid or nucleotide sequence.\n","\"\"\"\n","\n","def writeFasta(fileHandle, header, sequence, sequenceLineWidth=100):\n","  \"\"\" Function writes a fasta file to the given file handle.\n","  \"\"\"\n","  fileHandle.write('>' + header + \"\\n\") # Write the header line\n","  \n","  # Now write the sequence\n","  for i in range(0,len(sequence),sequenceLineWidth): # Step through sequence sequenceLineWidth characters\n","    # at a time\n","    \n","    # Write the next sequenceLineWidth chars of the sequence\n","    fileHandle.write(sequence[i:i+sequenceLineWidth] + \"\\n\") \n","    \n","def readFasta(fileHandle):\n","  \"\"\" Read a single sequence and header from a fasta file.\n","  \"\"\"\n","  \n","  # Read the header line, skipping and lines not beginning with a '>'\n","  while True: # Keep looping until we break\n","    l = fileHandle.readline()\n","    \n","    if len(l) == 0: # If we have reached the end of the file, terminate\n","      return None # None is a standard NULL or 0 quantity used to indicate\n","    # termination\n","    \n","    if l[0] == '>': # Is a valid header line\n","      header = l[1:-1] # Get the  header between the '>' and the '\\n'\n","      break\n","   \n","  # Now read the sequence\n","  substrings = []\n","  while True:\n","    i = fh.tell() # tell() gives the index that the file handle is at in the file\n","    \n","    l = fileHandle.readline()\n","    \n","    if len(l) == 0: # We have reached the end of the file\n","      break\n","    \n","    if l[0] == '>': # We have encountered the start of another sequence\n","      fh.seek(i) # Roll back the file handle to the point before we called readline()\n","      # this is like \"undoing\" the readline() call\n","      break\n","  \n","    substrings.append(l[:-1]) # Add the line, minus the newline to substrings\n","  \n","  return header, \"\".join(substrings)\n","  \n","# A few test sequences\n","dna_sequences = [ \n","    (\"a_dummy_dna_string\", \"GATTACA\"),\n","    (\"gi 556503834 ref NC_000913.3 Escherichia coli str. K-12 substr. MG1655, complete genome\", \"AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGCTTCTGAACTGGTTACCTGCCGTGAGTAAATTAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAATATAGGCATAGCGCACAGACAGATAAAAATTACAGAGTACACAACATCCATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCACAGGTAACGGTGCGGGCTGACGCGTACAGGAAACACAGAAAAAAGCCCGCACCTGACAGTGCGGGCTTTTTTTTTCGACCAAAGGTAACGAGGTAACAACCATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAACGTTTTCTGCGTGTTGCCGATATTCTGGAAAGCAATGCCAGGCAGGGGCAGGTGGCCACCGTCCTCTCTGCCCCCGCCAAAATCACCAACCACCTGGTGGCGATGATTGAAAAAACCATTAGCGGCCAGGATGCTTTACCCAATATCAGCGATGCCGAACGTATTTTTGCCGAACTTTTGACGGGACTCGCCGCCGCCCAGCCGGGGTTCCCGCTGGCGCAATTGAAAACTTTCGTCGATCAGGAATTTGCCCAAATAAAACATGTCCTGCATGGCATTAGTTTGTTGGGGCAGTGCCCGGATAGCATCAACGCTGCGC\"),\n","    (\"another_dummy_dna_string\", \"CGGCGCGCGTCTTTGCAGCGATGTCACGCGCCCGTATTTCCGTGGTGCTGATTACGCAATCATCTTCCGAATACAGCATCAGTTTCTGCGTTCCACAAAGCGACTGTGTGCGAGCTGAACGGGCAATGCAGGAAGAGTTC\"),]\n","\n","# Write the sequences into the file\n","with open(\"fasta.fa\", \"w\") as fh:\n","  for header, sequence in dna_sequences:\n","    writeFasta(fh, header, sequence)\n","    \n","# Let's take a quick look at the file we wrote:\n","with open(\"fasta.fa\") as fh:\n","  print(\"Printing fasta file\")\n","  for line in fh:\n","    print(line, end=\"\") # end=\"\" stops print() from adding a newline to the end\n","\n","# Now let's read them back\n","dna_sequences_copy = []\n","with open(\"fasta.fa\") as fh:\n","  while True:\n","    i = readFasta(fh)\n","    if i == None:\n","      break\n","    dna_sequences_copy.append(i)\n","    \n","# We can check it all worked\n","assert dna_sequences == dna_sequences_copy"],"execution_count":null,"outputs":[{"output_type":"stream","text":["Printing fasta file\n",">a_dummy_dna_string\n","GATTACA\n",">gi 556503834 ref NC_000913.3 Escherichia coli str. K-12 substr. MG1655, complete genome\n","AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGCTTCTGAACTGGTTACCTGCCGTGAGTAAAT\n","TAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAATATAGGCATAGCGCACAGACAGATAAAAATTACAGAGTACACAACATCCATGAAACGCAT\n","TAGCACCACCATTACCACCACCATCACCATTACCACAGGTAACGGTGCGGGCTGACGCGTACAGGAAACACAGAAAAAAGCCCGCACCTGACAGTGCGGG\n","CTTTTTTTTTCGACCAAAGGTAACGAGGTAACAACCATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAACGTTTTCTGCGTGTTGCCG\n","ATATTCTGGAAAGCAATGCCAGGCAGGGGCAGGTGGCCACCGTCCTCTCTGCCCCCGCCAAAATCACCAACCACCTGGTGGCGATGATTGAAAAAACCAT\n","TAGCGGCCAGGATGCTTTACCCAATATCAGCGATGCCGAACGTATTTTTGCCGAACTTTTGACGGGACTCGCCGCCGCCCAGCCGGGGTTCCCGCTGGCG\n","CAATTGAAAACTTTCGTCGATCAGGAATTTGCCCAAATAAAACATGTCCTGCATGGCATTAGTTTGTTGGGGCAGTGCCCGGATAGCATCAACGCTGCGC\n",">another_dummy_dna_string\n","CGGCGCGCGTCTTTGCAGCGATGTCACGCGCCCGTATTTCCGTGGTGCTGATTACGCAATCATCTTCCGAATACAGCATCAGTTTCTGCGTTCCACAAAG\n","CGACTGTGTGCGAGCTGAACGGGCAATGCAGGAAGAGTTC\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"h0iDQunV8V9W"},"source":["# Directories\n","\n","In most systems the operating has access to a file system in which files are kept. Getting into details\n","is beyond the scope of this course, because it requires us getting into operating system stuff - but let's cover the very basics.\n","\n","For more see: https://docs.python.org/3/library/os.html and https://docs.python.org/3/library/os.path.html\n","\n","By default when you open/write a file in Python uses the current working directory."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":34},"id":"sdevYDXf8XT6","executionInfo":{"elapsed":954,"status":"ok","timestamp":1572562873979,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"76fbef28-9349-4e4a-ae84-1fab16b66865"},"source":["import os # This module provides all sorts of useful functions \n","# for working with the operating system\n","\n","os.getcwd() # getcwd() tells us what working directory we're in. \n","# (If you're using Google Collaboratory, this is some directory on\n","# some (virtual) machine in Google Cloud Platform)"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'/content'"]},"metadata":{"tags":[]},"execution_count":18}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":34},"id":"y-ug5LCkjl4c","executionInfo":{"elapsed":901,"status":"ok","timestamp":1572562959521,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"759cc447-56c4-46b2-be7a-7c8f5299ecd0"},"source":["os.listdir(os.getcwd()) # listdir() tells us what files are in the working directory"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['.config', 'test.txt2', 'test.txt', 'test2.txt', 'fasta.fa', 'sample_data']"]},"metadata":{"tags":[]},"execution_count":19}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":34},"id":"Se2Rr5isj6I7","executionInfo":{"elapsed":845,"status":"ok","timestamp":1572563150292,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"b6014ba6-cd3f-4094-8b28-cc80ddc97724"},"source":["# We can create a directory like so:\n","\n","os.mkdir(\"new_directory\")\n","\n","os.listdir(os.getcwd() + \"/new_directory\") # Now \"new_directory\" is a directory in our working directory"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["[]"]},"metadata":{"tags":[]},"execution_count":25}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":34},"id":"3LTGLKkqkRVB","executionInfo":{"elapsed":623,"status":"ok","timestamp":1572563281536,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"eeb04894-a7e2-4ca6-94ed-69ac8cba96b2"},"source":["# We can write a file in our new directory like so:\n","\n","# Python uses strings to indicate file paths, generally we can indicate file\n","# paths with the / symbol, so new_directory/test.txt is a file \"test.txt\" in the \n","# directory \"new_directory\"\n","\n","with open(\"new_directory/test.txt\", \"w\") as fh:\n","  for header, sequence in dna_sequences: # Write the DNA sequences into the file\n","    writeFasta(fh, header, sequence)\n","    \n","    \n","os.listdir(os.getcwd() + \"/new_directory\") # Now \"new_directory\" is a directory in our working directory"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['test.txt']"]},"metadata":{"tags":[]},"execution_count":28}]},{"cell_type":"markdown","metadata":{"id":"v7VoOWRTp66n"},"source":["It is worth mentioning how to delete a file, because your Python programs/scripts should try not make lots of messy files:"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":34},"id":"Q8OHe-OKp9La","executionInfo":{"elapsed":491,"status":"ok","timestamp":1572563337020,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"1deb2c8d-fdf6-48ba-94f3-ad0286e7ca66"},"source":["os.remove(\"new_directory/test.txt\") # This removes the file test.txt from new_directory\n","\n","os.listdir(os.getcwd() + \"/new_directory\") # Now the directory is empty"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["[]"]},"metadata":{"tags":[]},"execution_count":30}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":34},"id":"ql-fl1mvqVwr","executionInfo":{"elapsed":678,"status":"ok","timestamp":1572563338388,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"33bf49bd-2eaf-4494-8848-3e1ff68b8670"},"source":["# Similarly you can remove empty directories\n","os.rmdir(\"new_directory\") # This will error out if the directory is not empty\n","\n","os.listdir(os.getcwd()) # new_directory is gone"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['.config', 'test.txt2', 'test.txt', 'test2.txt', 'fasta.fa', 'sample_data']"]},"metadata":{"tags":[]},"execution_count":31}]},{"cell_type":"markdown","metadata":{"id":"LeFOCs6H8dfE"},"source":["# Getting Data from the Internet"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":69},"id":"b-OAJZVF8iq7","executionInfo":{"elapsed":494,"status":"ok","timestamp":1572563503696,"user":{"displayName":"Benedict Paten","photoUrl":"https://lh3.googleusercontent.com/a-/AAuE7mBMnWy8dDR7jyTHNy9tPaRx6DCyA3QKrIcuQ7R4=s64","userId":"06399644931392855882"},"user_tz":420},"outputId":"6296a5bd-eb88-4ffa-e0c4-14f1de11ab57"},"source":["import urllib.request\n","\n","url = \"https://raw.githubusercontent.com/DataBiosphere/toil/master/src/toil/test/wdl/test.csv\"\n","local_copy = \"local.txt\"\n","\n","urllib.request.urlretrieve(url, local_copy) # This function copies the thing the url points at into\n","# a local file copy\n","\n","with open(local_copy) as fh: # Print the file\n","  for line in fh:\n","    print(line, end=\"\")"],"execution_count":null,"outputs":[{"output_type":"stream","text":["1,2,3\n","4,5,6\n","7,8,9\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"uWr3iUZ_WDYj"},"source":["# Challenge Example"]},{"cell_type":"code","metadata":{"id":"UgYsFrEKE4Y4"},"source":["# Complete the following code.\n","\n","l = list(range(100))\n","\n","## Write the numbers in l to a file \"out.txt\"\n","## Read the numbers from out.txt into a new list, \"l2\"\n","## write to file read from file\n","## check that l == l2"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"KZP9y5sgyVm8"},"source":["# Reading\n","\n","* Open book Chapter 13 on files: http://openbookproject.net/thinkcs/python/english3e/files.html\n","\n","# Homework\n","\n","* ZyBook Reading 11\n","\n"]}]}